Implement custom callout

nikischin opened the pull request:
Hi @Nicolapps,

I finally managed to put some effort in the custom callouts again. I was struggling a lot with the implementation honestly, though I quite like the current way I did. Let me know what you think!

There is still some things I need to do before merging:

- [x] Implement also for annotation
- [x] Cleanup code and documentation
- [x] Create pretty examples for storybooks with some nice styling and example data

Though it would be nice to have a first review on the current progress already, as this should be already fine to review.

Thank you so much! I would be really happy to replace my current svg workaround I build for my implementation with this new feature :)

Greetings Tim

Nicolapps reviewed (commented):
Hey Tim,

Thank you very much for your contribution, and sorry I couldn’t review it earlier! The new feature looks great and I’m looking forward to seeing it merged.

I added a few suggestions in my review. Are there still things you need to do on your side?

nikischin commented:
Hi @Nicolapps thank you so much for the great improvements. I am currently on holidays and won't have time to implement them yet but I have had a quick look at it already and it really looks reasonable!! :)) Thank you so much!

nikischin commented on src/components/Marker.tsx:
```diff
@@ -82,6 +103,64 @@ export default function Marker({
+      // @ts-ignore
+      delete marker.callout;
+    }
+  }, [
+    marker, calloutElementForAnnotation, calloutLeftAccessoryForAnnotation,
+    calloutRightAccessoryForAnnotation, calloutContentForAnnotation,
+    calloutElementForAnnotationRef.current, calloutLeftAccessoryForAnnotationRef.current,
+    calloutRightAccessoryForAnnotationRef.current, calloutContentForAnnotationRef.current,
+  ]);
```
Yeah, only did this because of the line limit...

nikischin commented on src/components/CalloutContainer.tsx:
```diff
@@ -0,0 +1,12 @@
+  { children, type = 'container' }: { children: ReactNode, type?: string },
+  ref,
+) => (
+  <div className={`mk-callout-${type}`} ref={ref}>
```
Ah, no I implemented it in order to provide a way to style the div introduced...

nikischin replied:
I need a `ref` of an html element which I hand over to MapKitJS. So this is why I implemented the `<div>` with the class name. Of course, the user can nest a custom `<div>` within this `<div>` but would have no easy way to access the outer `<div>` for styling.

nikischin commented on src/components/CalloutContainer.tsx:
```diff
@@ -0,0 +1,12 @@
+
+const CalloutContainer = React.forwardRef((
+  { children, type = 'container' }: { children: ReactNode, type?: string },
+  ref,
```
So I implemented your suggestion and also did this before, however, now I get TS errors in other places I wasn't able to fix. I am not so experienced in TS though, so maybe you could have a look at it.

nikischin commented on src/components/AnnotationProps.tsx:
```diff
@@ -142,5 +142,82 @@ export default interface AnnotationProps {
+   * annotation as a parameter.
+   * @see {@link https://developer.apple.com/documentation/mapkitjs/annotationcalloutdelegate/2991148-calloutcontentforannotation}
+   */
+  calloutElementForAnnotation?: JSX.Element;
```
I agree, though this is the same behavior as in mapkitJS. Generally speaking the Typescript solution preventing the setting of both of them would be the the cleanest I guess, I don't think we should implement an error, which mapkitJS wouldn't throw on it's own...

nikischin replied:
I tried to implement a conditional type, however, I don't really know how to do that with interfaces.

I was trying something like:

```
...Markerprops extends CalloutProps ...

//doesn't work: Cannot mix types and interfaces in that case
interface CalloutProps extends CalloutElemProps {
...calloutEnabled ...
}


type CalloutElemProps = CalloutElementProps | CalloutContentProps;

interface CalloutElementProps {
...calloutElement...
}

interface CalloutContentProps {
...calloutContent...
...calloutLeft...
...calloutRight...
}
```

Nicolapps replied:
I thought there would be a way to fix this by using `type` instead of `interface`. Something like:

```ts
export type MarkerProps = {
  // ...
  calloutEnabled?: boolean;
} & ({
  calloutLeftAccessory?: ReactNode;
  calloutRightAccessory?: ReactNode;
  calloutContent?: ReactNode;

  calloutElement?: never;
} | {
  calloutLeftAccessory?: never;
  calloutRightAccessory?: never;
  calloutContent?: never;

  calloutElement?: ReactNode;
});
```

But I’ve just tried it and while it produces the correct type (e.g. it only allows to set either calloutElement or the other properties at once), it hurts the developer experience (VS Code isn’t capable of showing the attribute documentation when using a union type for the props). So I think that it's best to keep using your current implementation.

nikischin replied:
Ok, should we still implement a error which is thrown if both attributes are set or is it rather not necessary?

Nicolapps replied:
No, I don’t think it is necessary since MapKit JS doesn’t do it.

nikischin commented:
Hi @Nicolapps I haven't yet have time to implement any example storybook for the annotation which definitely would be cool to have, but implemented or commented the other suggestions you had. 

Could you have a quick look at the comments and also let me know if I implemented your suggestions correctly. Most of your suggestion apply on both the marker and annotation component, I really hope I implemented all improvements on both of those components.

nikischin commented on src/components/AnnotationProps.tsx:
```diff
@@ -142,5 +142,82 @@ export default interface AnnotationProps {
+   * When MapKit JS creates a callout for a selected annotation and the annotation’s callout
+   * delegate has no calloutElementForAnnotation method, the framework calls
+   * calloutContentForAnnotation method instead — if it’s defined — on the delegate with the
+   * annotation as a parameter.
```
Could you come up with a suggestion for an improved description? Thank you!

nikischin commented:
Hi @Nicolapps 

So I implemented the storybook example for the annotation, could you check the implementation again and check on the open comments?

Thank you so much!

Tim

Nicolapps commented:
Hi Tim, thank you for your changes! I’m going to look at them later this week, as soon as I have some time.

Nicolapps commented on src/components/Annotation.tsx:
```diff
@@ -90,7 +172,11 @@ export default function Annotation({
     visible,
+
     clusteringIdentifier,
+    displayPriority,
```
You should make sure you handle the new string values for `displayPriority` correctly (a correction function in `parameters.ts` could be helpful.

nikischin replied:
What do you mean by an "correction function"? Like ensuring it's between 0 and 1000 but then we are not talking about string values... also ts should check that it's a number (not necessarily within range) or one of the allowed strings.

nikischin replied:
So I had a look into `parameters.ts` and think you are asking to introduce enum values instead of the current string behavior maybe? So we would have enums for `low`, `high` and `required`? Could be a good idea, I agree!

Nicolapps replied:
Sorry, I meant conversion, not correction. We don’t need to do parameter validation here, but when the user sets `"high"` as the parameter value we should send the value `mapkit.Annotation.DisplayPriority.High` to MapKit JS (and do the same for the two other string values). I don’t think an enum is necessary here.

Nicolapps reviewed (commented):
Thank you a lot for these changes!

nikischin commented:
Thank you so much again for your valuable review! I implemented the requested changes and commented on the two other conversations (the classname for the div and the conditional attributes are still open). Let me know what you think about those. Thank you so much!

Nicolapps reviewed (approved):
Thank you a lot for your PR!

Nicolapps commented:
The new feature is released in v1.15.0! 🚀
